Add `localized` method for querysets

tbrlpld opened the pull request:
<!-- Thanks for contributing to Wagtail! 🎉  Please add a description below, explaining the purpose of this pull request - including the issue number of the issue you're fixing (if applicable). -->

Fixes #...







_Please check the following:_

-   [ ] Do the tests still pass?[^1]
-   [ ] Does the code comply with the style guide?
    -   [ ] Run `make lint` from the Wagtail root.
-   [ ] For Python changes: Have you added tests to cover the new/fixed behaviour?
-   [ ] For front-end changes: Did you test on all of Wagtail’s supported environments?[^2]
    -   [ ] **Please list the exact browser and operating system versions you tested**:
    -   [ ] **Please list which assistive technologies [^3] you tested**:
-   [ ] For new features: Has the documentation been updated accordingly?

**Please describe additional details for testing this change**.

[^1]: [Development Testing](https://docs.wagtail.org/en/latest/contributing/developing.html#testing)
[^2]: [Browser and device support](https://docs.wagtail.org/en/latest/contributing/developing.html#browser-and-device-support)
[^3]: [Accessibility Target](https://docs.wagtail.org/en/latest/contributing/developing.html#accessibility-targets)

ababic commented on wagtail/models/i18n.py:
```diff
@@ -174,13 +174,143 @@ def is_active(self) -> bool:
+            # Keep the same order as in the original queryset. To do so, we annotate the
+            # localized queryset with the original order of the translation keys, and
+            # then order by that annotation.
+            ordering_when_clauses = [
+                models.When(translation_key=tk, then=models.Value(index))
+                for index, tk in enumerate(original_translation_keys)
+            ]
+            localized_annotated_queryset = localized_queryset.annotate(
+                original_order=models.Case(*ordering_when_clauses)
+            )
```
This is 🤌!

tbrlpld replied:
Heavily inspired by your `order_by_pk_position` util in Kit 😅

ababic commented on wagtail/models/i18n.py:
```diff
@@ -174,13 +174,143 @@ def is_active(self) -> bool:
+        # Get all instances that are not available in the active locale. We can find
+        # these by excluding the translation keys for which translations exist from the
+        # original queryset.
+        translated_translation_keys = translated_instances.values_list(
+            "translation_key", flat=True
+        )
+        untranslated_instances = self.exclude(
+            translation_key__in=translated_translation_keys,
+        )
+
+        # Combine the two querysets to get the localized queryset.
+        localized_queryset = self.model.objects.filter(
+            models.Q(pk__in=translated_instances)
+            | models.Q(pk__in=untranslated_instances)
+        )
```
```suggestion
        translated_vals = translated_instances.values_list(
            "translation_key", "pk"
        )
        translated_translation_keys = (key for key, pk in translated_vals)
        translated_pks = (pk for key, pk in translated_vals)
        untranslated_pks = self.exclude(
            translation_key__in=translated_translation_keys,
        ).values_list("pk", flat=True)

        # Combine the two querysets to get the localized queryset.
        localized_queryset = self.model.objects.filter(
            models.Q(pk__in=translated_pks)
            | models.Q(pk__in=untranslated_pks)
        )
```

ababic replied:
`translated_vals` in the above suggestion fetches the keys and pks in a single query, meaning `localized_queryset` doesn't need to make another

I've also updated the `untranslated_pks` query to specifically fetch pks, as the entire queryset would have been used before `localised_queryset`

tbrlpld replied:
I am not sure I follow. The entire localized queryset is already fetched in a single query. The keys and the separate querysets are never actually fetched. They are only used to build the final query. 

I actually think that iterating over the values list would trigger an additional query no? 
```python
        translated_translation_keys = (key for key, pk in translated_vals)
        translated_pks = (pk for key, pk in translated_vals)
```

ababic replied:
Wow. Django's lazy queryset evaluation truly is on another level these days! It'll be a complicated query for sure, but a single database round-trip, which makes all the difference. Excellelent work @tbrlpld!

tbrlpld replied:
Thanks @ababic. 

Yea, it's pretty amazing what the resolver can do. I looked at the query once and can definitely confirm it's on the bigger side 😅

ababic commented on wagtail/models/i18n.py:
```diff
@@ -174,13 +174,143 @@ def is_active(self) -> bool:
+        """
+        Return a localized version of this queryset.
+
+        Localize means to check, for every instance in the queryset, if there is a
+        translation into the active locale. If a translation exists, use the translated
+        instance, otherwise use the instance itself. The result is a queryset of the
+        same length as the original queryset.
```
A localized queryset is one where objects are replaced (where possible) with versions translated into the active locale.

ababic replied:
Suggestion:
```suggestion
           A localized queryset is one where objects are replaced with versions translated into the active locale.
           Where a translation isn't available for an object, the original is returned instead.
```

ababic commented on wagtail/models/i18n.py:
```diff
@@ -174,13 +174,143 @@ def is_active(self) -> bool:
+        # behavior by passing `include_draft_translations=True`.
+        from wagtail.models import DraftStateMixin
+
+        if issubclass(self.model, DraftStateMixin) and not include_draft_translations:
+            translated_instances = translated_instances.exclude(live=False)
```
I think the handling of this is sensible!

ababic commented on wagtail/models/i18n.py:
```diff
@@ -174,13 +174,143 @@ def is_active(self) -> bool:
+        instance, otherwise use the instance itself. The result is a queryset of the
+        same length as the original queryset.
+
+        By default, the same ordering definition as in the original queryset is applied.
+        This means that the translated values are being considered during ordering,
+        which can lead to a different order than the original queryset. If you want to
+        preserve the same order as the original queryset, you need to pass
+        ``keep_order=True``.
```
It's probably also worth mentioning that any annotations applied to the original queryset will be lost as part of the process. We can advise that developers apply annotations after calling `localized()` on it.

ababic commented on wagtail/models/i18n.py:
```diff
@@ -174,13 +174,143 @@ def is_active(self) -> bool:
+
+    def localized(
+        self: models.QuerySet,
+        keep_order: bool = False,
```
```suggestion
        preserve_order: bool = False,
```

tbrlpld replied:
That does sound better 😅

tbrlpld replied:
509afb27612828e7c8b4dc92d1b0e8dcd133ec09

tbrlpld commented on wagtail/models/i18n.py:
```diff
@@ -174,13 +174,143 @@ def is_active(self) -> bool:
+    def localized(
+        self: models.QuerySet,
+        keep_order: bool = False,
+        include_draft_translations: bool = False,
```
a8d9b4f73a0b7b636444ffa92a64881b2639e239

tbrlpld commented:
Thanks for the review @ababic! I have incorporated most of your suggestions and would be happy to hear your feedback on those changes. 

I did not yet include the PK suggestion, but I am happy to discuss that one further.

jhonatan-lopes commented on wagtail/models/i18n.py:
```diff
@@ -174,13 +174,182 @@ def is_active(self) -> bool:
+        # queryset and that are available in the active locale.
+        active_locale = Locale.get_active()
+        original_translation_keys = self.values_list("translation_key", flat=True)
+        translated_instances = self.model.objects.filter(
```
Sorry to drop in unannounced, but shouldn't references to the model manager be made to the default manager? I.e. to `_default_manager` and not to `objects`, in case they are not the same:

```
translated_instances = self.model._default_manager.filter(
```

tbrlpld replied:
Good shout. I followed what I have seen other places, but I am happy to give this a go.

jhonatan-lopes commented:
FYI, I was trying to port this implementation to Mozilla before it officially comes into Wagtail to get the `preserve_order` behaviour and I ran into some issues. We've had a similar [`localize_queryset` utility function](https://github.com/MozillaFoundation/foundation.mozilla.org/blob/5288b376f98cceb39bedf1ffbde5990ed89094f7/network-api/networkapi/wagtailpages/utils.py#L325-L363) in the codebase for a while but it lacks the ability to keep the original order of the queryset.

After porting over this PR's implementation, this piece of code started to give me problems:

```python
        top_topics = (
            BlogPageTopic.objects.filter(blogpage__authors__author=author_profile)
            .annotate(count=models.Count("name"))
            .order_by("-count")[:3]
        )

        top_topics = localize_queryset(top_topics)

        print(top_topics)
```
with a `django.core.exceptions.FieldError: Cannot resolve expression type, unknown output_field` error message.

I believe this is caused by the `count` annotation getting removed by the localisation queryset. Then, when `localize_queryset` (or `.localized` in this PR) tries to order the queryset by the original order (`localized_queryset.order_by(*queryset.query.order_by)`), it tries to access `count` which doesn't exist on `localized_queryset`. If I remove the annotation and the `order_by`, the code functions normally.

I'm wrecking my brain trying to rewrite the query so that I can localize before but I don't think that's possible, since we need to count the topics on foreign key relationships that only exist in the default locale.

My next step is to try to keep the original annotations, which wasn't an issue with our current implementation.

I think at least providing a check that the field exists with a proper exception before trying to reorder would be helpful, because the error message here is no help at all...

ababic commented:
@jhonatan-lopes wouldn't you just apply the annotation and ordering AFTER? e.g.:

```
top_topics = (
            BlogPageTopic.objects.filter(blogpage__authors__author=author_profile)
            .localize()
            .annotate(count=models.Count("name"))
            .order_by("-count")[:3]
        )

```

ababic commented:
@tbrlpld Have we documented that annotations are currently lost? A better option might be to preserve them in the final query by copying from `self.query.annotations`.

jhonatan-lopes commented:
Hey @ababic, I'm afraid the count annotation (and consequently the ordering) will be wrong if I annotate after, since they will try to count the number of times the translated topic has been referenced -- which will be zero, since we always keep track of just the default locale one, i.e. the relationships on the CMS are defined based on the default locale and then localised as needed on views.

tbrlpld commented:
> @tbrlpld Have we documented that annotations are currently lost? A better option might be to preserve them in the final query by copying from `self.query.annotations`.

@ababic Yea, it is in the docstring now that the annotations are getting lost. I do like the idea of copying them over from the original though. What other options are there? I would guess things like `select_related`, `prefetch_related` etc. are also lost.

tbrlpld commented:
> Hey @ababic, I'm afraid the count annotation (and consequently the ordering) will be wrong if I annotate after, since they will try to count the number of times the translated topic has been referenced -- which will be zero, since we always keep track of just the default locale one, i.e. the relationships on the CMS are defined based on the default locale and then localised as needed on views.

Thanks for checking it out @jhonatan-lopes! I will try to make annotations work, as I see this would be useful. 

In the meantime, one other option for you could be to save the order of the `translation_keys` and then apply the order with `When` clauses to the localized queryset. 

This will probably need another query but I guess that could be acceptable. 

```python 
top_topics = (
    BlogPageTopic.objects.filter(blogpage__authors__author=author_profile)
)

localized_top_topics = localize_queryset(top_topics)

ordered_top_topics_translation_keys = (
    top_topics
    .annotate(count=models.Count("name"))
    .order_by("-count")[:3]
    .values_list("translations_keys", flat=True)
)

ordering_when_clauses = [
    models.When(translation_key=tk, then=models.Value(index))
     for index, tk in enumerate(ordered_top_topics_translation_keys)
]
# Annotate localized queryset with the order and order by it
localized_and_ordered_top_topics = (
    localized_top_topics
    .annotate(original_order=models.Case(*ordering_when_clauses))
    .order_by("original_order)
)
```

jhonatan-lopes commented:
Thanks @tbrlpld! I'm also taking a look at adding/copying annotations, as I need it to work on Moz as well. I'll post here if I make some progress

jhonatan-lopes commented on wagtail/models/i18n.py:
```diff
@@ -174,13 +174,182 @@ def is_active(self) -> bool:
+                models.Q(pk__in=translated_instances)
+                | models.Q(pk__in=untranslated_instances)
+            )
+
```
Hey @tbrlpld, this works to get the annotations back into the localized queryset:

```suggestion
if annotations := queryset.query.annotations:
    localized_queryset = localized_queryset.annotate(**annotations)

```

However, the values will be based on the localized queryset, so that the annotations will be updated on the translated instances. Maybe it's worth highlighting that.

The good thing is that it definitely allows us to retain the original order of the queryset when the ordering was done through an annotation.

tbrlpld replied:
> However, the values will be based on the localized queryset, so that the annotations will be updated on the translated instances. Maybe it's worth highlighting that.

That should be expected, no? Annotations are make on a per-row basis, so a different instance should be expected to have a different value.

tbrlpld replied:
I guess your case would still require preserving the original ordering.

tbrlpld replied:
Thanks @jhonatan-lopes. I have incorporated that. I had to make some extra additions to also make it work for [aliases](https://docs.djangoproject.com/en/4.2/ref/models/querysets#django.db.models.query.QuerySet.alias).

tbrlpld commented on wagtail/models/i18n.py:
```diff
@@ -174,13 +174,182 @@ def is_active(self) -> bool:
+            localized_annotated_queryset = localized_queryset.annotate(
+                original_order=models.Case(*ordering_when_clauses)
+            )
+            return localized_annotated_queryset.order_by("original_order")
```
Added. Thanks @jhonatan-lopes

tbrlpld commented:
Just wanted to bump this for review, as it has been a while.

zerolab commented:
@tbrlpld any chance you can rebase it? We'll add it to the queue

tbrlpld commented:
Will do @zerolab

tbrlpld commented on wagtail/tests/test_management_commands.py:
The query counts needed to be bumped by 1. 

    
The query count depends on the number of models that inherit from
`DraftStateMixin`.

For testing of the interaction of localization and draft states, a
`TestDraftModel` is added.

This new model now needs to be retrieved in the tests for the
scheduled publishing, thus increasing the query count by one.

tbrlpld commented:
@zerolab I have rebased the PR and fixed a small issue with the query count due to the `TestDraftModel`. 

Let me know if this needs extra changes to move this forward.
